src/main.rs

```diff
@@ -42,6 +50,41 @@ struct Cli {
     )]
     groups: Vec<String>,
 }
+pub struct EmitKVPLayer {
+    file_path: std::path::PathBuf,
+}
+
+impl EmitKVPLayer {
+    pub fn new(file_path: std::path::PathBuf) -> Self {
+        EmitKVPLayer { file_path }
+    }
+}
+
+impl<S> Layer<S> for EmitKVPLayer
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup>,
+{
+    fn on_new_span(
+        &self,
+        _attrs: &Attributes<'_>,
+        id: &Id,
+        ctx: TracingContext<'_, S>,
+    ) {
+        let start_time = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_millis();
+        if let Some(span) = ctx.span(id) {
+            span.extensions_mut().insert(start_time);
```

Comment:
Okay, so there are a few things here including a gotcha to how extensions work.

## Instant
First, there's a neat type available to calculate durations without needing to touch `SystemTime` at all called [Instant](https://doc.rust-lang.org/std/time/struct.Instant.html). If you create an `Instant`, you can later query how long it's been with:
```Rust
let then = std::time::Instant::now();

// A Duration with the time between `then` and `now`
let now = now.elapsed();
```

You can also compare between instants with `Instant::duration_since` if you have two instants and want to know the time between them, but you shouldn't need that here.

## Extensions
Extensions are span-specific, which means layers from other libraries (OpenTelemetry, the layers that come with tracing-subscriber, some other custom layer we might add later) can see and access the extension. This is mostly not a problem, with one gotcha: the extension entry is looked up by its _type_.

This means that if another layer wants to insert an extension that is the same type as the type we add here, it'll cause the program to panic. `start_time` is `u128` so if any other layer tries to insert a `u128` we'll crash. There's a line at the end of the [insert](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/registry/struct.ExtensionsMut.html#method.insert) API doc talks a bit about this, but if you're not already familiar with the pattern it is probably confusing. It says that extensions should generally be "newtypes".

## Newtype pattern

The newtype pattern is when you make a new type to wrap some other type. This is useful in scenarios where you want to provide some additional functionality on top of an existing type, or in cases like this where we really want to make sure no other layer could provide a conflicting type. It looks like this when applied to the standard library's `Instant` type.

```Rust
// Define a new tuple struct with a single member, the type we're wrapping.
struct MyInstant(std::time::Instant);

fn main() {
    let then = MyInstant(std::time::Instant::now());
    let now = then.0.elapsed();
}
```

The neat thing about this is that because we don't mark `MyInstant` as `pub`, we can be certain no other library could possibly insert the same type into the span's extensions since we're the only crate with access.

### Deref

This is not necessary to do, but it's nice to know. It can be a little annoying to do the `instant.0.elapsed()` thing, so one thing we can do is implement [Deref](https://doc.rust-lang.org/std/ops/trait.Deref.html) for our new type:

```Rust
struct MyInstant(std::time::Instant);

impl std::ops::Deref for MyInstant {
    type Target = std::time::Instant;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

fn main() {
    let then = MyInstant(std::time::Instant::now());
    let now = then.elapsed();
}
```
The compiler will silently insert calls to our `deref` implementation so we can treat references to our new type as references to the inner type. This is neat, but as the documentation on the trait mentions, it's something [you want to be careful with](https://doc.rust-lang.org/std/ops/trait.Deref.html#when-to-implement-deref-or-derefmut). In this case it's okay, though, since it's a private type and we're implementing no additional methods.

## The Fix

All this to say, you'll want to use `Instant` instead of converting to a time since the epoch, and you'll want to wrap it in a new type. You'll need to do some re-arranging of the code to make that work, since currently `handle_span` is in a different crate (and since we want to keep our new type private, the `azurekvp` crate won't be able to access it).

My recommendation would be to move the contents of the `azurekvp` crate into `src/kvp.rs` so it's part of the `azure-init` crate. You'll need to add a `mod kvp;` to `src/main.rs`.

If we later want to provide a generic, sharable implementation of a tracing layer for KVP we can break it out into its own crate, but for now I don't think we should split them up.